### ISSUE TITLE
Agora eventually blocks, runs out of memory and dies under load

### ISSUE THREAD

**Issue by Geod24** [bosagora/agora#1480]
```
Jan 10 13:20:01 na-002 docker[11697]: 2021-01-10 13:19:57,949 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:14 na-002 docker[11697]: 2021-01-10 13:20:14,781 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:27 na-002 docker[11697]: 2021-01-10 13:20:24,595 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:46 na-002 systemd[1]: agora@7.service: Main process exited, code=exited, status=137/n/a
Jan 10 13:55:46 na-002 systemd[1]: agora@7.service: Failed with result 'exit-code'.
Jan 10 13:55:46 na-002 systemd[1]: agora@7.service: Service hold-off time over, scheduling restart.
Jan 10 13:55:46 na-002 systemd[1]: agora@7.service: Scheduled restart job, restart counter is at 1.
Jan 10 13:55:46 na-002 systemd[1]: Stopped Agora node 7.
Jan 10 13:55:46 na-002 systemd[1]: Started Agora node 7.
Jan 10 13:55:55 na-002 docker[4948]: 2021-01-10 13:55:55,240 Info [agora.node.TransactionPool] - Loading database from: .cache/tx_pool.dat
Jan 10 13:55:55 na-002 docker[4948]: 2021-01-10 13:55:55,260 Info [agora.node.TransactionPool] - Loaded database from: .cache/tx_pool.dat
Jan 10 13:55:55 na-002 docker[4948]: 2021-01-10 13:55:55,425 Info [agora.consensus.state.UTXODB] - Loading UTXO database from: .cache/utxo_set.dat
Jan 10 13:55:55 na-002 docker[4948]: 2021-01-10 13:55:55,425 Info [agora.consensus.state.UTXODB] - Loaded database from: .cache/utxo_set.dat
```

The alerts in the live system turned out to be quite useful. From time to time, some instances become unresponsive.
This happens while we have Faucet running and flooding the network. The GC usage goes through the roof (currently it's 15.5G and nothing is happening) and memory ends up being exhausted. We should figure out where our memory leak comes from.

**Comment by Geod24** [comment]
Note that the other instance on the server didn't crash like this, but was unresponsive for quite a bit:
```
Jan 10 13:19:25 na-002 docker[11514]: Error while writing the response: Connection closed while writing data.
Jan 10 13:19:36 na-002 docker[11514]: 2021-01-10 13:19:33,59 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Payment
Jan 10 13:19:41 na-002 docker[11514]: 2021-01-10 13:19:33,205 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:19:49 na-002 docker[11514]: 2021-01-10 13:19:47,147 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:09 na-002 docker[11514]: 2021-01-10 13:20:08,325 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:20 na-002 docker[11514]: 2021-01-10 13:20:20,258 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:23 na-002 docker[11514]: 2021-01-10 13:20:20,878 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:20:26 na-002 docker[11514]: 2021-01-10 13:20:21,180 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:21 na-002 docker[11514]: 2021-01-10 13:55:17,224 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:21 na-002 docker[11514]: 2021-01-10 13:55:18,370 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:22 na-002 docker[11514]: 2021-01-10 13:55:19,184 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:22 na-002 docker[11514]: 2021-01-10 13:55:20,112 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:22 na-002 docker[11514]: 2021-01-10 13:55:21,146 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:22 na-002 docker[11514]: Failed to get remote address for TCP connection
Jan 10 13:55:23 na-002 docker[11514]: 2021-01-10 13:55:22,87 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Payment
Jan 10 13:55:23 na-002 docker[11514]: Failed to get remote address for TCP connection
Jan 10 13:55:23 na-002 docker[11514]: 2021-01-10 13:55:23,147 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:23 na-002 docker[11514]: Failed to get remote address for TCP connection
Jan 10 13:55:25 na-002 docker[11514]: 2021-01-10 13:55:25,390 Fatal [agora.consensus.protocol.Nominator] - tryNominate(): Invalid consensus data: Enrollment: Insufficient number of active validators. Data: { tx_set: [{ type: TxType.Paymen
Jan 10 13:55:25 na-002 docker[11514]: Failed to get remote address for TCP connection
```

And the dmesg of the machine shows it was indeed the OOM killer:
```
[Jan10 13:59] accounts-daemon invoked oom-killer: gfp_mask=0x14200ca(GFP_HIGHUSER_MOVABLE), nodemask=(null), order=0, oom_score_adj=0
[  +0.000001] accounts-daemon cpuset=/ mems_allowed=0
[  +0.000005] CPU: 7 PID: 22055 Comm: accounts-daemon Tainted: G          I      4.15.0-118-generic #119-Ubuntu
[  +0.000001] Hardware name: Supermicro X8STi/X8STi, BIOS 2.0        09/17/10
[...]
[  +0.000001] Mem-Info:
[  +0.000005] active_anon:7234967 inactive_anon:654765 isolated_anon:0
               active_file:203 inactive_file:126 isolated_file:0
               unevictable:0 dirty:0 writeback:0 unstable:0
               slab_reclaimable:64641 slab_unreclaimable:163403
               mapped:266 shmem:50 pagetables:20740 bounce:0
               free:49465 free_pcp:829 free_cma:0
[  +0.000004] Node 0 active_anon:28939868kB inactive_anon:2619060kB active_file:812kB inactive_file:504kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:1064kB dirty:0kB writeback:0kB shmem:200kB shmem_thp: 0kB shmem_pmdmapp
[  +0.000001] Node 0 DMA free:15884kB min:32kB low:44kB high:56kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15968kB managed:15884kB mlocked:0kB kernel_stack:0kB pagetables
[  +0.000004] lowmem_reserve[]: 0 3439 32080 32080 32080
[  +0.000003] Node 0 DMA32 free:121776kB min:7240kB low:10760kB high:14280kB active_anon:3417368kB inactive_anon:4268kB active_file:124kB inactive_file:352kB unevictable:0kB writepending:0kB present:3644928kB managed:3579360kB mlocked:0kB
[  +0.000004] lowmem_reserve[]: 0 0 28640 28640 28640
[  +0.000002] Node 0 Normal free:60200kB min:60308kB low:89636kB high:118964kB active_anon:25522500kB inactive_anon:2614792kB active_file:768kB inactive_file:568kB unevictable:0kB writepending:0kB present:29884416kB managed:29334568kB mlo
[  +0.000004] lowmem_reserve[]: 0 0 0 0 0
[  +0.000002] Node 0 DMA: 1*4kB (U) 1*8kB (U) 0*16kB 0*32kB 2*64kB (U) 1*128kB (U) 1*256kB (U) 0*512kB 1*1024kB (U) 1*2048kB (M) 3*4096kB (M) = 15884kB
[  +0.000009] Node 0 DMA32: 2303*4kB (UME) 747*8kB (UME) 109*16kB (UME) 101*32kB (UME) 50*64kB (UME) 15*128kB (ME) 6*256kB (ME) 1*512kB (M) 1*1024kB (M) 0*2048kB 23*4096kB (UME) = 122564kB
[  +0.000008] Node 0 Normal: 1915*4kB (ME) 6561*8kB (UMEH) 1*16kB (H) 0*32kB 0*64kB 0*128kB 0*256kB 1*512kB (H) 1*1024kB (H) 0*2048kB 0*4096kB = 61700kB
[  +0.000009] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[  +0.000000] 917 total pagecache pages
[  +0.000002] 344 pages in swap cache
[  +0.000001] Swap cache stats: add 539519, delete 539173, find 271037/273848
[  +0.000001] Free swap  = 0kB
[  +0.000000] Total swap = 1046520kB
[  +0.000001] 8386328 pages RAM
[  +0.000001] 0 pages HighMem/MovableOnly
[  +0.000001] 153875 pages reserved
[  +0.000000] 0 pages cma reserved
[  +0.000001] 0 pages hwpoisoned
[...]
[  +0.000001] Out of memory: Kill process 11810 (agora) score 481 or sacrifice child
[  +0.092786] Killed process 11810 (agora) total-vm:18111016kB, anon-rss:15875196kB, file-rss:0kB, shmem-rss:0kB
[  +0.915554] oom_reaper: reaped process 11810 (agora), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

Sweet 18G of VM used.

**Comment by Geod24** [comment]
After a lot of trial and error, and discussions with @ferencdg and @kchulj , it seems to come from the fact that threads are initialized and shut down very frequently. I think that's an issue in Vibe.d. Fixing #1143 would fix the issue, but threads start / stop are expensive regardless and this shouldn't happen. PR #1577 has been very helpful in giving us insights and I'll expand it further.

**Comment by kchulj** [comment]
Which part of [#1143](https://github.com/bpfkorea/agora/pull/1143) would need to be fixed?

**Comment by Geod24** [comment]
Typo, I meant #1443

**Comment by Geod24** [comment]
While debugging with @kchulj we had an epiphany: We never free memory allocated by `std::vector`. And we call `toVec` a lot, especially since the nodes are stuck nominating most of the time, e.g. here:
https://github.com/bpfkorea/agora/blob/b878017ea1124ac81e81b003371e2bb319cc4212/source/agora/consensus/protocol/Nominator.d#L421-L422

**Comment by omerfirmak** [comment]
Root cause: https://github.com/vibe-d/eventcore/issues/178

**Comment by AndrejMitrovic** [comment]
> Root cause: vibe-d/eventcore#178

Nice find. I thought that code looked dodgy when I first saw it.

**Comment by Geod24** [comment]
Omer to the rescue. Fixed by #1852 .
